File: sdks/react-native/ios/OmiModule.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(OmiModule, RCTEventEmitter)

RCT_EXTERN_METHOD(connect:(NSString *)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(disconnect:(NSString *)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isConnected:(NSString *)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getAudioCodec:(NSString *)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(startAudioBytesNotifications:(NSString *)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopAudioBytesNotifications:(NSString *)deviceId
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(startScan:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(stopScan:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject)

@end

File: sdks/react-native/android/src/main/java/com/omi/reactnative/OmiModule.java
package com.omi.reactnative;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothManager;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class OmiModule extends ReactContextBaseJavaModule {
    private static final String TAG = "OmiModule";
    
    // Service and characteristic UUIDs
    private static final UUID OMI_SERVICE_UUID = UUID.fromString("19b10000-e8f2-537e-4f6c-d104768a1214");
    private static final UUID AUDIO_DATA_STREAM_CHARACTERISTIC_UUID = UUID.fromString("19b10001-e8f2-537e-4f6c-d104768a1214");
    private static final UUID AUDIO_CODEC_CHARACTERISTIC_UUID = UUID.fromString("19b10002-e8f2-537e-4f6c-d104768a1214");
    private static final UUID BUTTON_SERVICE_UUID = UUID.fromString("23ba7924-0000-1000-7450-346eac492e92");
    private static final UUID BUTTON_TRIGGER_CHARACTERISTIC_UUID = UUID.fromString("23ba7925-0000-1000-7450-346eac492e92");
    private static final UUID BATTERY_SERVICE_UUID = UUID.fromString("0000180f-0000-1000-8000-00805f9b34fb");
    private static final UUID BATTERY_LEVEL_CHARACTERISTIC_UUID = UUID.fromString("00002a19-0000-1000-8000-00805f9b34fb");
    
    private final ReactApplicationContext reactContext;
    private BluetoothAdapter bluetoothAdapter;
    private final Map<String, BluetoothGatt> connectedDevices = new HashMap<>();
    
    public OmiModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        
        BluetoothManager bluetoothManager = (BluetoothManager) reactContext.getSystemService(Context.BLUETOOTH_SERVICE);
        if (bluetoothManager != null) {
            bluetoothAdapter = bluetoothManager.getAdapter();
        }
    }
    
    @NonNull
    @Override
    public String getName() {
        return "OmiModule";
    }
    
    private void sendEvent(String eventName, @Nullable WritableMap params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }
    
    @ReactMethod
    public void connect(String deviceId, Promise promise) {
        // Mock implementation for now
        promise.resolve(null);
    }
    
    @ReactMethod
    public void disconnect(String deviceId, Promise promise) {
        // Mock implementation for now
        promise.resolve(null);
    }
    
    @ReactMethod
    public void isConnected(String deviceId, Promise promise) {
        // Mock implementation for now
        promise.resolve(false);
    }
    
    @ReactMethod
    public void getAudioCodec(String deviceId, Promise promise) {
        // Mock implementation for now - return PCM8 (1)
        promise.resolve(1);
    }
    
    @ReactMethod
    public void startAudioBytesNotifications(String deviceId, Promise promise) {
        // Mock implementation for now
        promise.resolve(null);
    }
    
    @ReactMethod
    public void stopAudioBytesNotifications(String deviceId, Promise promise) {
        // Mock implementation for now
        promise.resolve(null);
    }
    
    @ReactMethod
    public void startScan(Promise promise) {
        // Mock implementation for now
        promise.resolve(null);
    }
    
    @ReactMethod
    public void stopScan(Promise promise) {
        // Mock implementation for now
        promise.resolve(null);
    }
    
    @ReactMethod
    public void addListener(String eventName) {
        // Keep: Required for RN built in Event Emitter Calls
    }
    
    @ReactMethod
    public void removeListeners(Integer count) {
        // Keep: Required for RN built in Event Emitter Calls
    }
}
